src/core/dbus-manager.c

```diff
@@ -2937,6 +2937,168 @@ static int method_dump_unit_descriptor_store(sd_bus_message *message, void *user
         return method_generic_unit_operation(message, userdata, error, bus_service_method_dump_file_descriptor_store, 0);
 }
 
+static int aux_scope_from_message(Manager *m, sd_bus_message *message, Unit **ret_scope, sd_bus_error *error) {
+        _cleanup_(sd_bus_creds_unrefp) sd_bus_creds *creds = NULL;
+        _cleanup_free_ pid_t *pids = NULL;
+        const char *name, *suffix;
+        Unit *from, *scope;
+        CGroupContext *cc;
+        size_t n_pids = 0;
+        pid_t pid, main;
+        uint64_t flags;
+        int r;
+
+        assert(ret_scope);
+
+        r = sd_bus_query_sender_creds(message, SD_BUS_CREDS_PID, &creds);
+        if (r < 0)
+                return r;
+
+        r = sd_bus_creds_get_pid(creds, &pid);
+        if (r < 0)
+                return r;
+
+        from = manager_get_unit_by_pid(m, pid);
+        if (!from)
+                return sd_bus_error_set(error, BUS_ERROR_NO_SUCH_UNIT, "Client not member of any unit.");
+
+        if (from->type != UNIT_SERVICE)
+                return sd_bus_error_setf(error, SD_BUS_ERROR_INVALID_ARGS,
+                                         "Starting auxiliary scope is supported only for service units, unit '%s' is not a service, refusing.", from->id);
+
+        if (!unit_name_is_valid(from->id, UNIT_NAME_PLAIN))
+                return sd_bus_error_setf(error, SD_BUS_ERROR_INVALID_ARGS,
+                                         "Auxiliary scope can be started only for non-template service units, refusing.");
+
+        r = sd_bus_message_read(message, "s", &name);
+        if (r < 0)
+                return r;
+
+        if (!unit_name_is_valid(name, UNIT_NAME_PLAIN))
+                return sd_bus_error_setf(error, SD_BUS_ERROR_INVALID_ARGS,
+                                         "Invalid name \"%s\" for auxiliary scope.", name);
+
+        suffix = strrchr(name, '.');
+        assert(suffix);
+
+        if (unit_type_from_string(suffix+1) != UNIT_SCOPE)
+                return sd_bus_error_setf(error, SD_BUS_ERROR_INVALID_ARGS,
+                                         "Name \"%s\" of auxiliary scope doesn't have .scope suffix.", name);
+
+        main = unit_main_pid(from);
+
+        r = sd_bus_message_enter_container(message, 'a', "h");
+        if (r < 0)
+                return r;
+
+        for (;;) {
+                _cleanup_free_ char *unit = NULL;
+                int fd;
+                pid_t q;
+
+                r = sd_bus_message_read(message, "h", &fd);
+                if (r < 0)
+                        return r;
+                if (r == 0)
+                        break;
+
+                r = pidfd_get_pid(fd, &q);
```

Comment:
I wonder if parsing here and below should be best-effort? IE, if something doesn't match - the process already exited, etc, then simply skip that process, but continue iterating. This would make it a bit more robust, for example you don't need to worry about a worker process existing in the time you have sent the request and are waiting for a response.

Reply:
Hmm, maybe something like this would work better

```C
r = pidfd_get_pid(fd, &q);
if (r < 0) {
        log_error_errno(r, "Failed to convert PIDFD to PID, ignoring: %m");
        continue;
}

r = cg_pid_get_unit(q, &unit);
if (r < 0) {
        log_error_errno(r, "Failed to obtain unit for PID " PID_FMT ", ignoring: %m", q);
        continue;
}
```

Reply:
yeah, and same for the main pid check